rust/pgrust/src/connection/flow.rs

```diff
@@ -0,0 +1,1234 @@
+    }
+}
+
+impl<'a> Flow for BindFlow<'a> {
+    fn to_vec(&self) -> Vec<u8> {
+        let mut format_codes = Vec::with_capacity(self.params.len());
+        let mut values = Vec::with_capacity(self.params.len());
+
+        for param in self.params {
+            match param {
+                Param::Null => {
+                    format_codes.push(0);
+                    values.push(Encoded::Null);
+                }
+                Param::Text(value) => {
+                    format_codes.push(0);
+                    values.push(Encoded::Value(value.as_bytes()));
+                }
+                Param::Binary(value) => {
+                    format_codes.push(1);
+                    values.push(Encoded::Value(value));
+                }
+            }
+        }
+
+        let result_format_codes = bytemuck::cast_slice(self.result_format_codes);
+
+        builder::Bind {
+            portal: self.portal.0,
+            statement: self.statement.0,
+            format_codes: &format_codes,
+            values: &values,
+            result_format_codes,
+        }
+        .to_vec()
+    }
+}
+
+impl<'a> Flow for ExecuteFlow<'a> {
+    fn to_vec(&self) -> Vec<u8> {
+        let max_rows = match self.max_rows {
+            MaxRows::Unlimited => 0,
+            MaxRows::Limited(n) => n.get() as i32,
+        };
+        builder::Execute {
+            portal: self.portal.0,
+            max_rows,
+        }
+        .to_vec()
+    }
+}
+
+impl<'a> Flow for DescribePortalFlow<'a> {
+    fn to_vec(&self) -> Vec<u8> {
+        builder::Describe {
+            name: self.name.0,
+            dtype: b'P',
+        }
+        .to_vec()
+    }
+}
+
+impl<'a> Flow for DescribeStatementFlow<'a> {
+    fn to_vec(&self) -> Vec<u8> {
+        builder::Describe {
+            name: self.name.0,
+            dtype: b'S',
+        }
+        .to_vec()
+    }
+}
+
+impl<'a> Flow for ClosePortalFlow<'a> {
+    fn to_vec(&self) -> Vec<u8> {
+        builder::Close {
+            name: self.name.0,
+            ctype: b'P',
+        }
+        .to_vec()
+    }
+}
+
+impl<'a> Flow for CloseStatementFlow<'a> {
+    fn to_vec(&self) -> Vec<u8> {
+        builder::Close {
+            name: self.name.0,
+            ctype: b'S',
+        }
+        .to_vec()
+    }
+}
+
+impl<'a> Flow for QueryFlow<'a> {
+    fn to_vec(&self) -> Vec<u8> {
+        builder::Query { query: self.query }.to_vec()
+    }
+}
+
+pub(crate) enum MessageResult {
+    Continue,
+    Done,
+    SkipUntilSync,
+    Unknown,
+    UnexpectedState { complaint: &'static str },
+}
+
+pub(crate) trait MessageHandler {
+    fn handle(&mut self, message: Message) -> MessageResult;
+    fn name(&self) -> &'static str;
+    fn is_sync(&self) -> bool {
+        false
+    }
+}
+
+pub(crate) struct SyncMessageHandler;
+
+impl MessageHandler for SyncMessageHandler {
+    fn handle(&mut self, message: Message) -> MessageResult {
+        if ReadyForQuery::try_new(&message).is_some() {
```

Comment:
Take out the `ReadyForQuery.status` and store in the connection? Or even return as a result of `pipeline_sync()`